### PR TITLE
Move function/actor exporting & loading code to function_manager.py

### PR DESCRIPTION
**Pull request by guoyuhong**
## What do these changes do?
Move function/actor exporting & loading code to function_manager.py to prepare the code change for function descriptor for python.

## Related issue number
https://github.com/ray-project/ray/issues/2576

### CODE REVIEW

**Comment by robertnishihara** [comment]
@suquark can you take a look at this?

**Comment by robertnishihara** [comment]
Thanks @guoyuhong! Can you say if there are any behavior changes or if it is only code reorganization?

**Comment by suquark** [comment]
@robertnishihara  I think it is similiar to https://github.com/ray-project/ray/pull/2602 where `python/ray/execution_info.py` resembles `function_manager.py`. In my full implementation https://github.com/ray-project/ray/pull/2577 there are also things dealing with actors & workers. But they are pending because we have not yet reached a consensus.

**Comment by suquark** [comment]
So my suggestion is that we should be more decisive in refactoring. We should first discuss when to start refactoring (because we are still building a new release) and define a time window in which to complete the refactoring tasks.

**Comment by robertnishihara** [comment]
Thanks @suquark, this PR feels fairly self-contained to me, so while I agree that major refactoring to worker.py should probably be put off until 0.6 is released, incremental cleanups that can be done quickly are probably a good idea.

**Comment by robertnishihara** [comment]
@guoyuhong we're tracking tasks for 0.6 in https://github.com/ray-project/ray/projects/5. The **To do** column is the highest priority stuff and the **Optional backlog** column is optional.

Probably somewhere between the middle of October and the end of October.

**Comment by guoyuhong** [comment]
@robertnishihara Thanks for the clarification. Will this PR be accepted right now or only after Release0.6?

**Comment by robertnishihara** [comment]
I think it should be merged before. I’ll try to review it later today or
tomorrow. I’ll discuss with @suquark.

File: python/ray/actor.py [hunk 1]
```diff
@@ -134,32 +107,9 @@ def set_actor_checkpoint(worker, actor_id, checkpoint_index, checkpoint,
...
-    if checkpoint_index is not None:
-        checkpoint_index = int(checkpoint_index)
-    return checkpoint_index, checkpoint, frontier
-
-
 def save_and_log_checkpoint(worker, actor):
     """Save a checkpoint on the actor and log any errors.
-
-    Args:
+     Args:
```
Comment (robertnishihara) [inline python/ray/actor.py]: newline before `Args:`

File: python/ray/actor.py [hunk 2]
```diff
@@ -182,8 +132,7 @@ def save_and_log_checkpoint(worker, actor):
 
 def restore_and_log_checkpoint(worker, actor):
     """Restore an actor from a checkpoint and log any errors.
-
-    Args:
+     Args:
```
Comment (robertnishihara) [inline python/ray/actor.py]: newline before `Args:`

File: python/ray/function_manager.py [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+import hashlib
+import inspect
+import json
+import time
+import traceback
+from collections import (
+    namedtuple,
+    defaultdict,
+)
+import ray
```
Comment (robertnishihara) [inline python/ray/function_manager.py]: newline before ray imports

File: python/ray/function_manager.py [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+
+
+FunctionExecutionInfo = namedtuple("FunctionExecutionInfo",
+                                   ["function", "function_name", "max_calls"])
+"""FunctionExecutionInfo: A named tuple storing remote function information."""
+
+
+class FunctionManager(object):
+    """
+    A class used to export/load remote functions.
```
Comment (robertnishihara) [inline python/ray/function_manager.py]: Should be

```python
"""A class used to export/load remote functions.
```

File: python/ray/function_manager.py [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+
+    Attributes:
+
```
Comment (robertnishihara) [inline python/ray/function_manager.py]: Please add the documentation

File: python/ray/function_manager.py [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+            self._functions_to_export.append(remote_function)
+            return
+        if self._worker.mode != ray.worker.SCRIPT_MODE:
+            # Don't need to export if the worker is not a driver.
+            return
+        self._do_export(remote_function)
+
+    def _do_export(self, remote_function):
+        """
+        Pickle a remote function and export it to redis.
```
Comment (robertnishihara) [inline python/ray/function_manager.py]: Move to previous line

File: python/ray/function_manager.py [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+    A class used to export/load remote functions.
+
+    Attributes:
+
+    """
+
+    def __init__(self, worker):
+        self._worker = worker
+        self._functions_to_export = []
+        self._actor_to_export = []
```
Comment (robertnishihara) [inline python/ray/function_manager.py]: probably should be `_actors_to_export`.

File: python/ray/function_manager.py [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+                    function=function,
+                    function_name=function_name,
+                    max_calls=max_calls))
+            # Add the function to the function table.
+            self._worker.redis_client.rpush(
+                b"FunctionTable:" + function_id.id(), self._worker.worker_id)
+
+    def get_execution_info(self, driver_id, function_id):
+        """
+        Get the FunctionExecutionInfo of a remote function.
```
Comment (robertnishihara) [inline python/ray/function_manager.py]: Move to previous line

File: python/ray/function_manager.py [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+
+        Args:
+            driver_id: id of the driver that the function belongs to.
+            function_id: id of the function to get.
```
Comment (robertnishihara) [inline python/ray/function_manager.py]: `id` -> `ID`, in both lines above

File: python/ray/function_manager.py [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+                        ray.utils.push_error_to_driver(
+                            self._worker,
+                            ray_constants.WAIT_FOR_FUNCTION_PUSH_ERROR,
+                            warning_message,
+                            driver_id=driver_id)
+                    warning_sent = True
+            time.sleep(0.001)
+
+    @classmethod
+    def function_predictor(cls, x):
```
Comment (robertnishihara) [inline python/ray/function_manager.py]: Maybe `is_function_or_method`?

**Review by robertnishihara — COMMENTED** [review]
Left very minor comments. I guess the main question is whether this should be a "FunctionManager" which managers remote functions and actors, or whether we should generalize it to handle all importing/exporting of Python functions (e.g., remote functions, actors, and "functions to run" (custom serializers)).

The more general approach is what @suquark did in a different PR, and I think that's probably the direction we want to go in (though probably in a follow up PR).

**Comment by guoyuhong** [comment]
@robertnishihara Thanks for the comments and communicating with @suquark . Yes, @suquark 's change may be a better choice for long term planning. Will his change be merged only after Release 0.6? Current PR is important to us because we want to enable calling Python code from Java as soon as possible due to the schedule.

File: python/ray/function_manager.py [hunk 11]
```diff
@@ -0,0 +1,495 @@
...
+                        ray.utils.push_error_to_driver(
+                            self._worker,
+                            ray_constants.WAIT_FOR_FUNCTION_PUSH_ERROR,
+                            warning_message,
+                            driver_id=driver_id)
+                    warning_sent = True
+            time.sleep(0.001)
+
+    @classmethod
+    def is_function_or_method(cls, x):
```
Comment (suquark) [inline python/ray/function_manager.py]: Could we put `is_function_or_method` &  `is_classmethod` into `ray.utils`? They are more like general type inspecting tools.

File: python/ray/function_manager.py [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+             ])
+
+        class_name = decode(class_name)
+        module = decode(module)
+        checkpoint_interval = int(checkpoint_interval)
+        actor_method_names = json.loads(decode(actor_method_names))
+
+        # Create a temporary actor with some temporary methods so that if
+        # the actor fails to be unpickled, the temporary actor can be used
+        # (just to produceerror messages and to prevent the driver from
```
Comment (suquark) [inline python/ray/function_manager.py]: A space between 'produceerror'.

**Comment by suquark** [comment]
This PR looks good for me. This could also be helpful for further worker refactoring.

**Comment by robertnishihara** [comment]
Thanks @suquark! I went ahead and made this changes. Will merge assuming tests pass.

**Review by robertnishihara — APPROVED** [review]

**Comment by guoyuhong** [comment]
@robertnishihara @suquark Thanks!

**Pull request merged**